streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogTopics.java

```diff
@@ -34,83 +40,64 @@
         this.internalTopicManager = internalTopicManager;
-        this.topicGroups = topicGroups;
-        this.tasksForTopicGroup = tasksForTopicGroup;
         final LogContext logContext = new LogContext(logPrefix);
         log = logContext.logger(getClass());
+        changelogTopicConfigs = computeChangelogTopicConfig(topicGroups, tasksForTopicGroup);
+        validationResult = internalTopicManager.validate(changelogTopicConfigs);
     }
 
     public void setup() {
-        // add tasks to state change log topic subscribers
-        final Map<String, InternalTopicConfig> changelogTopicMetadata = new HashMap<>();
-        for (final Map.Entry<Integer, TopicsInfo> entry : topicGroups.entrySet()) {
-            final int topicGroupId = entry.getKey();
-            final TopicsInfo topicsInfo = entry.getValue();
-
-            final Set<TaskId> topicGroupTasks = tasksForTopicGroup.get(topicGroupId);
-            if (topicGroupTasks == null) {
-                log.debug("No tasks found for topic group {}", topicGroupId);
-                continue;
-            } else if (topicsInfo.stateChangelogTopics.isEmpty()) {
-                continue;
-            }
-
-            for (final TaskId task : topicGroupTasks) {
-                final Set<TopicPartition> changelogTopicPartitions = topicsInfo.stateChangelogTopics
-                    .keySet()
-                    .stream()
-                    .map(topic -> new TopicPartition(topic, task.partition))
-                    .collect(Collectors.toSet());
-                changelogPartitionsForStatefulTask.put(task, changelogTopicPartitions);
-            }
-
-            for (final InternalTopicConfig topicConfig : topicsInfo.nonSourceChangelogTopics()) {
-                // the expected number of partitions is the max value of TaskId.partition + 1
-                int numPartitions = UNKNOWN;
-                for (final TaskId task : topicGroupTasks) {
-                    if (numPartitions < task.partition + 1) {
-                        numPartitions = task.partition + 1;
-                    }
-                }
-                topicConfig.setNumberOfPartitions(numPartitions);
-                changelogTopicMetadata.put(topicConfig.name(), topicConfig);
-            }
-            sourceTopicBasedChangelogTopics.addAll(topicsInfo.sourceTopicChangelogs());
+        if (!validationResult.misconfigurationsForTopics().isEmpty()) {
+            throw new MisconfiguredInternalTopicException(Utils.join(misconfigured().values().stream()
+                .flatMap(Collection::stream).collect(Collectors.toList()), Utils.NL)
+            );
         }
 
-        final Set<String> newlyCreatedChangelogTopics = internalTopicManager.makeReady(changelogTopicMetadata);
-        log.debug("Created state changelog topics {} from the parsed topology.", changelogTopicMetadata.values());
+        final Set<String> missingTopics = validationResult.missingTopics();
```

Comment:
So I understand correctly - In a future change we'll actually not create the missing topics if not configured to do so?

Reply:
Yes, exactly. Before missing topics would have been created implicitly. Now we first validate the topics to discover missing topics and misconfigurations and then according to a config we decide if we create the missing topics or if we fail.